sdk/lib/web3.js

```diff
@@ -136,20 +129,58 @@ async function singleCall({ web3, limiter, target, abi, block, params }) {
   return { callCount: 1, output: result };
 }
 
-async function multiCall({ web3, limiter, target, abi, block, calls }) {
+async function multiCall({ multiCallProvider, limiter, target, abi, block, calls }) {
   debug("multiCall", target, abi.name, block, calls);
 
-  const rateLimitedCallOne = limiter.wrap(callOne);
-
-  const results = await Promise.all(calls.map(arg => rateLimitedCallOne(web3, abi, arg.target, arg.params)));
-
-  return { callCount: calls.length, output: results };
+  if (calls.length === 0) return [];
+  const templateContract = new ethers.Contract(calls[0]?.target ?? target ?? "", [abi], multiCallProvider);
+  const callChunks = _.chunk(calls, 400);
+
+  const rateLimitedCallChunk = limiter.wrap(async (callChunk) => {
+    return await Promise.allSettled(
+      callChunk.map((call) => {
+        const contract = templateContract.attach(call.target ?? target ?? "");
+        const params = call.params !== undefined ? (Array.isArray(call.params) ? call.params : [call.params]) : [];
+        const resultPromise = contract[abi.name ?? ""](...params, {
+          blockTag: block,
+        });
+        return resultPromise;
+      })
+    );
+  });
```

Comment:
multiCall function becomes long, maybe callChunk can be extracted to be a separated function to make the main logic shorter